aws-lc-sys/builder/cc_builder.rs

```diff
@@ -0,0 +1,297 @@
+            .flag("-Wno-unused-parameter")
+            .cpp(false)
+            .shared_flag(false)
+            .static_flag(true);
+        if target_os() == "linux" {
+            cc_build.define("_XOPEN_SOURCE", "700").flag("-lpthread");
+        }
+        if let Some(prefix) = &self.build_prefix {
+            cc_build
+                .define("BORINGSSL_IMPLEMENTATION", "1")
+                .define("BORINGSSL_PREFIX", prefix.as_str());
+        }
+        self.add_includes(&mut cc_build);
+        cc_build
+    }
+
+    fn add_includes(&self, cc_build: &mut cc::Build) {
+        cc_build
+            .include(self.manifest_dir.join("include"))
+            .include(self.manifest_dir.join("generated-include"))
+            .include(self.manifest_dir.join("aws-lc").join("include"))
+            .include(
+                self.manifest_dir
+                    .join("aws-lc")
+                    .join("third_party")
+                    .join("s2n-bignum")
+                    .join("include"),
+            );
+    }
+
+    fn add_all_files(&self, lib: &Library, cc_build: &mut cc::Build) {
+        use core::str::FromStr;
+        cc_build.file(PathBuf::from_str("rust_wrapper.c").unwrap());
+
+        for source in lib.sources {
+            let source_path = self.manifest_dir.join("aws-lc").join(source);
+            let is_asm = std::path::Path::new(source)
+                .extension()
+                .map_or(false, |ext| ext.eq("S"));
+            if is_asm && target_vendor() == "apple" && target_arch() == "aarch64" {
+                let mut cc_preprocessor = self.create_builder();
+                cc_preprocessor.file(source_path);
+                let preprocessed_asm = String::from_utf8(cc_preprocessor.expand()).unwrap();
+                let preprocessed_asm = preprocessed_asm.replace(';', "\n\t");
+                let asm_output_path = self.out_dir.join(source);
+                fs::create_dir_all(asm_output_path.parent().unwrap()).unwrap();
+                fs::write(asm_output_path.clone(), preprocessed_asm).unwrap();
+                cc_build.file(asm_output_path);
+            } else {
+                cc_build.file(source_path);
+            }
+        }
+    }
+
+    fn build_library(&self, lib: &Library) {
+        let mut cc_build = self.create_builder();
+
+        self.add_all_files(lib, &mut cc_build);
+
+        for flag in lib.flags {
+            cc_build.flag(flag);
+        }
+        self.compiler_checks(&mut cc_build);
+
+        if let Some(prefix) = &self.build_prefix {
+            cc_build.compile(format!("{}_crypto", prefix.as_str()).as_str());
+        } else {
+            cc_build.compile(lib.name);
+        }
+    }
+
+    // This performs basic checks of compiler capabilities and sets an appropriate flag on success.
+    // This should be kept in alignment with the checks performed by AWS-LC's CMake build.
+    // See: https://github.com/search?q=repo%3Aaws%2Faws-lc%20check_compiler&type=code
```

Comment:
Not required for this PR but something to think about long term: what mechanism do we need to ensure these compiler checks stay in sync with the AWS-LC CMake build? In s2n-tls we went with the approach of scanning the features directory in each build system, which automatically picks up new features (or removes unused ones). Maybe we should do something similar here?

https://github.com/aws/s2n-tls/pull/3921